src/compiler/factory/emitHelpers.ts

```diff
@@ -1367,6 +1387,66 @@ export const classPrivateFieldInHelper: UnscopedEmitHelper = {
             };`
 };
 
+/**
+ * @internal
+ */
+export const addDisposableResourceHelper: UnscopedEmitHelper = {
+    name: "typescript:addDisposableResource",
+    importName: "__addDisposableResource",
+    scoped: false,
+    text: `
+        var __addDisposableResource = (this && this.__addDisposableResource) || function (env, value, async) {
+            if (value !== null && value !== void 0) {
+                if (typeof value !== "object") throw new TypeError("Object expected.");
+                var dispose;
+                if (async) {
+                    if (!Symbol.asyncDispose) throw new TypeError("Symbol.asyncDispose is not defined.");
+                    dispose = value[Symbol.asyncDispose];
+                }
+                if (dispose === void 0) {
+                    if (!Symbol.dispose) throw new TypeError("Symbol.dispose is not defined.");
+                    dispose = value[Symbol.dispose];

```

Comment:
In Babel we use `Symbol.for("Symbol.dispose")` and `Symbol.for("Symbol.asyncDispose")` as fallbacks, so that users can use polyfills that do not modify the global scope. This is very useful for libraries, so that they can do something
```js
class MyResource {
  [Symbol.dispose || Symbol.for("Symbol.dispose")]() {}
}
```
and they will be usable both natively and in older environments.

It might be great to align on this.

Reply:
TypeScript helpers generally don't polyfill/shim Symbols in this way. We usually depend on the developer to introduce any necessary global shim instead. I'm curious if @DanielRosenwasser has any thoughts on this approach, however.

Reply:
We discussed this in design meeting and don't believe TypeScript should support `Symbol.for()` as a fallback. For one, `Symbol.for("Symbol.dispose")` is only marginally an improvement over just using a string like `"@@dispose"`, and we don't treat the result of `Symbol.for()` as a meaningfully unique symbol, which is a requirement for us to properly type check it as a property name. As a result, a property name of `[Symbol.dispose || Symbol.for("Symbol.dispose")]` produces a compilation error, so you wouldn't be able to produce that code without turning off error checking. Even if we did allow `Symbol.for()` to produce a meaningfully unique symbol that we can type check, we would essentially have to bifurcate `Disposable` to support each method name, and deal with the complexity of that only working in older compilation targets, since the `Symbol.for()` fallback wouldn't be supported natively in `--target esnext`.

Babel supporting for `Symbol.for()` is actually a bit of a hazard for TypeScript. If someone were to publish a package that leveraged Babel's support for a `Symbol.for()` fallback, they might in turn indicate that their objects are "disposable" in their documentation. That could lead to someone else putting together types for that package on DefinitelyTyped that indicate a class has a `Symbol.dispose` method when it does not. This in turn could lead to errors during type check if the package's consumer isn't running with `--lib esnext.disposable`, which could then result in them changing their `--lib` to make the error go away without ensuring that their runtime environment _actually_ supports a global `Symbol.dispose`.

While I think it is commendable to support users' efforts to avoid global scope modifications, I don't think this is an approach TypeScript can take here. I'd even go so far as to recommend that Babel not provide such support, but given that Babel's support for `Symbol.for("Symbol.dispose")` only has an _exteremly_ remote possibility of causing problems for TypeScript projects, I'll leave that up to you.

I'll also note that, since the native `DisposableStack` has a built-in mechanism to adopt non-disposable resources via `stack.adopt(value, v => v.close())` and `stack.defer(() => { value.close(); })`, a workaround like `Symbol.for()` may not even be warranted.